docs/source/advanced/contrib_thunder.rst

```diff
@@ -0,0 +1,468 @@
+The *lowest* level is the primitive operations, defined in `thunder/core/prims.py <https://github.com/Lightning-AI/lightning-thunder/blob/main/thunder/core/prims.py>`_.
+Primitive operations, as seen in the `Representing operations <../advanced/inside_thunder.rst#representing-operations>`__ section,
+describe all the computation performed, and they are intended to be as simple as possible so that
+executors like `nvFuser <https://github.com/NVIDIA/Fuser>`_ find it easy to manipulate them.
+*Thunder's* primitives are similar to `PyTorch’s <https://pytorch.org>`_
+`primTorch primitives <https://github.com/pytorch/pytorch/tree/main/torch/_prims>`_,
+and are based on `JAX’s <https://github.com/google/jax>`_
+`jax.lax operations <https://jax.readthedocs.io/en/latest/jax.lax.html>`_.
+
+Primitives have several parts, as defined by the ``make_prim`` function.
+Most importantly they have an ``id``, a ``name``, and a ``meta`` function.
+The meta function performs error checking and maps the metadata (like ``dtype``, ``shape``, ``device``) of inputs to the primitive metadata of outputs.
+For operations that are part of a class, like the elementwise unary or reduction operations, they often share a common meta function.
+More unique operations, like ``slice``, define their own
+`meta functions <https://github.com/Lightning-AI/lightning-thunder/blob/888b46324462fba70f93d5017bc0d99025f05091/thunder/core/prims.py#L2812>`_.
+
+The actual execution of primitive operations is handled by executors like
+`nvFuser <https://github.com/NVIDIA/Fuser>`_ or
+`PyTorch <https://pytorch.org>`_ – more on that in a moment.
+
+Before adding a primitive, check with the team on its design.
+It is appropriate to add primitives when necessary to describe the semantics of an operation or to improve the numerical accuracy or speed of operations.
+
+There is a tension with the design of primitive operations.
+On one hand, fewer primitive operations can make program transformation, and execution easier.
+Fewer primitives means fewer transformation rules – since transformation rules are defined on primitives – and a smaller interface with executors.
+On the other hand, too few primitive operations may make it hard, or impossible, to express all the operations that practitioners are interested in.
+Too few primitive operations may also make it difficult to execute programs quickly and numerically accurately.
+
+For example, the ``expm1`` operation can mathematically be defined in terms of the ``exp`` and subtraction operations,
+and so it does not need to be a primitive to enable any functionality.
+Many libraries, including `C++’s standard library <https://en.cppreference.com/w/c/numeric/math/expm1>`_,
+still define an ``expm1`` operation for numerical accuracy, and so ``expm1`` is a
+`primitive <https://github.com/Lightning-AI/lightning-thunder/blob/888b46324462fba70f93d5017bc0d99025f05091/thunder/core/prims.py#L1791>`_ in *Thunder*.
+
+
+-----------------
+The Core Language
```

Comment:
This section describes core langauge very well, but it is missing describing the motivation for having this extra layer instead of merging it into prims. (I don't know what the motivation is, otherwise I would make a more precise suggestion)

Reply:
But the motivation is there, nor? Prims are simple, do not do promotions/broadcasting and such. Their goal, is, ideally, just compute. But we are not very consistent with that, to be honest. We do have some quite powerful prims.

Reply:
Having the core language as a more powerful set of operators makes sense if the other higher-level languages (torch, ...) decompose into clang ops, not just prims.
Is that the case? I was under the impression that this isn't the case.

This is why I never understood the existence of the clang. Is it in case people want to write code with these operators directly instead of torch?

Reply:
Well, we can have high-level functions which makes our lives easier and not bound to the torch interface.

Reply:
Let's mention that here then, that is valid motivation

Reply:
I did re-read the sections. I believe they explain things well enough already. So, we have simple prims, and we have the torch lang that is used directly to be mapped into when seeing a torch op. Clang and prims are used to make that map happen. This is it there is to it. All other degrees of freedom are following from the "definitions".